macros/edr/data_monitoring/monitors_query/dimension_monitoring_query.sql

```diff
@@ -17,75 +25,93 @@
-                and {{ metric_properties.where_expression }}
-            {% endif %}
         ),
 
-        {# Outdated dimension values are dimensions with all metrics of 0 in the range of the test time #}
-        dimension_values_without_outdated as (
-            select distinct 
+        all_dimension_metrics as (
+            select
+                bucket_end,
                 dimension_value,
-                sum(metric_value)
+                metric_value,
+                row_number () over (partition by dimension_value order by bucket_end desc) as row_number
             from {{ ref('data_monitoring_metrics') }}
             where full_table_name = {{ full_table_name_str }}
                 and metric_name = {{ elementary.edr_quote(metric_name) }}
-                and dimension = {{ elementary.edr_quote(dimensions_string) }}
-                and {{ elementary.edr_cast_as_timestamp('bucket_end') }} >= {{ elementary.edr_cast_as_timestamp(min_bucket_start) }}
+                and metric_properties = {{ elementary.dict_to_quoted_json(metric_properties) }}
+        ),
+
+        training_set_dimensions as (
+            select distinct
+                dimension_value,
+                min(bucket_end) as dimension_min_bucket_end,
+                sum(metric_value)
+            from all_dimension_metrics
+            where row_number <= {{ test_configuration.min_training_set_size }}
             group by 1
+            {# Remove outdated dimension values (dimensions with all metrics of 0 in the range of the test time) #}
             having sum(metric_value) > 0
         ),
 
-        dimension_values_union as (
-            select distinct *
-            from (
-                select distinct 
-                    dimension_value,
-                    1 as joiner
-                from dimension_values_without_outdated
-                union all
-                select distinct
-                    dimension_value,
-                    1 as joiner
-                from filtered_monitored_table
-            ) results
+        unique_previous_dimension_values as (
+            select distinct
+                dimension_value,
+                dimension_min_bucket_end,
+                1 as joiner
+            from training_set_dimensions
         ),
 
-        {# Created buckets for each dimension value #}
+        {# Create buckets for each previous dimension value #}
         dimensions_buckets as (
             select edr_bucket_start, edr_bucket_end, dimension_value
-            from buckets left join dimension_values_union on buckets.joiner = dimension_values_union.joiner
+            from unique_previous_dimension_values left join buckets
+                on (buckets.joiner = unique_previous_dimension_values.joiner
+                {# This makes sure we don't create empty buckets for dimensions before their first apperance #}
+                and edr_bucket_end >= dimension_min_bucket_end)
```

Comment:
Like!
Can we add a test case for it in our e2e?

Reply:
Added, a bit hacky but @elongl is refactoring the e2e